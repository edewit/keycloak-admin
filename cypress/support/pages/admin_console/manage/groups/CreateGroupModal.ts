/// <reference types="../../../.." />

export default class CreateGroupModal {
  private nameInput = "groupNameInput";
  private createButton = "createGroup";

  fillGroupForm(name = "") {
    cy.getBy(this.nameInput).clear();
    cy.getBy(this.nameInput).type(name);
    return this;
  }

  clickCreate() {
    cy.getBy(this.createButton).click();
    return this;
  }
}
