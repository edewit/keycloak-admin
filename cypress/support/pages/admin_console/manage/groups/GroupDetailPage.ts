const expect = chai.expect;

export default class GroupDetailPage {
  private groupNamesColumn = '[data-label="Group name"] > a';
  private memberTab = "members";
  private attributesTab = "attributes";
  private memberNameColumn = 'tbody > tr > [data-label="Name"]';
  private includeSubGroupsCheck = "includeSubGroupsCheck";
  private addMembers = "addMember";
  private addMember = "add";
  private memberUsernameColumn = 'tbody > tr > [data-label="Username"]';

  private keyInput = '[name="attributes[0].key"]';
  private valueInput = '[name="attributes[0].value"]';
  private saveAttributeBtn = ".pf-c-form__actions > .pf-m-primary";

  checkListSubGroup(subGroups: string[]) {
    cy.get(this.groupNamesColumn).should((groups) => {
      expect(groups).to.have.length(subGroups.length);
      for (let index = 0; index < subGroups.length; index++) {
        const subGroup = subGroups[index];
        expect(groups).to.contain(subGroup);
      }
    });
    return this;
  }

  clickMembersTab() {
    cy.getId(this.memberTab).click();
    return this;
  }

  checkListMembers(members: string[]) {
    cy.get(this.memberNameColumn).should((member) => {
      expect(member).to.have.length(members.length);
      for (let index = 0; index < members.length; index++) {
        expect(member.eq(index)).to.contain(members[index]);
      }
    });
    return this;
  }

  checkSelectableMembers(members: string[]) {
    cy.get(this.memberUsernameColumn).should((member) => {
      for (const user of members) {
        expect(member).to.contain(user);
      }
    });
    return this;
  }

  selectUsers(users: string[]) {
    for (const user of users) {
      cy.get(this.memberUsernameColumn)
        .contains(user)
        .parent()
        .find("input")
        .click();
    }
    return this;
  }

  clickAdd() {
    cy.getId(this.addMember).click();
    return this;
  }

  clickIncludeSubGroups() {
    cy.getId(this.includeSubGroupsCheck).click();
    return this;
  }

  clickAttributesTab() {
    cy.getId(this.attributesTab).click();
    return this;
  }

  clickAddMembers() {
    cy.getId(this.addMembers).click();
    return this;
  }

  fillAttribute(key: string, value: string) {
    cy.get(this.keyInput).type(key).get(this.valueInput).type(value);
    return this;
  }

  saveAttribute() {
    cy.get(this.saveAttributeBtn).click();
    return this;
  }
}
