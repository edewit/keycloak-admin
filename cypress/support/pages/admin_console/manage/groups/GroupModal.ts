export default class GroupModal {
  private openButton = "openCreateGroupModal";
  private nameInput = "groupNameInput";
  private createButton = "createGroup";
  private renameButton = "renameGroup";

  open(name?: string) {
    cy.getId(name || this.openButton).click();
    return this;
  }

  fillGroupForm(name = "") {
    cy.getId(this.nameInput).clear().type(name);
    return this;
  }

  clickCreate() {
    cy.intercept("/auth/admin/realms/master/groups/*/members").as(
      "groupCreate"
    );
    cy.getId(this.createButton).click();
    cy.wait(["@groupCreate"]);

    return this;
  }

  clickRename() {
    cy.getId(this.renameButton).click();
    return this;
  }
}
