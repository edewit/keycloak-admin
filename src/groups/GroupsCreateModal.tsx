import React from "react";
import {
  AlertVariant,
  Button,
  Form,
  FormGroup,
  Modal,
  ModalVariant,
  TextInput,
  ValidatedOptions,
} from "@patternfly/react-core";
import { useTranslation } from "react-i18next";
import { useForm } from "react-hook-form";

import GroupRepresentation from "keycloak-admin/lib/defs/groupRepresentation";
import { useAdminClient } from "../context/auth/AdminClient";
import { useAlerts } from "../components/alert/Alerts";

type GroupsCreateModalProps = {
  id?: string;
  handleModalToggle: () => void;
  refresh: () => void;
};

export const GroupsCreateModal = ({
  id,
  handleModalToggle,
  refresh,
}: GroupsCreateModalProps) => {
  const { t } = useTranslation("groups");
  const adminClient = useAdminClient();
  const { addAlert } = useAlerts();
  const { register, errors, handleSubmit } = useForm();

  const submitForm = async (group: GroupRepresentation) => {
    try {
      if (!id) {
        await adminClient.groups.create({ name: group.name });
      } else {
        await adminClient.groups.setOrCreateChild({ id }, { name: group.name });
      }

      refresh();
      handleModalToggle();
      addAlert(t("groupCreated"), AlertVariant.success);
    } catch (error) {
      addAlert(t("couldNotCreateGroup", { error }), AlertVariant.danger);
    }
  };

  return (
    <>
      <Modal
        variant={ModalVariant.small}
        title={t("createAGroup")}
        isOpen={true}
        onClose={handleModalToggle}
        actions={[
          <Button
            data-testid="createGroup"
            key="confirm"
            variant="primary"
            type="submit"
            form="group-form"
          >
            {t("create")}
          </Button>,
        ]}
      >
        <Form id="group-form" isHorizontal onSubmit={handleSubmit(submitForm)}>
          <FormGroup
            name="create-modal-group"
            label={t("common:name")}
            fieldId="group-id"
            helperTextInvalid={t("common:required")}
            validated={
              errors.name ? ValidatedOptions.error : ValidatedOptions.default
            }
            isRequired
          >
            <TextInput
              data-testid="groupNameInput"
              ref={register({ required: true })}
              autoFocus
              type="text"
              id="create-group-name"
              name="name"
              validated={
                errors.name ? ValidatedOptions.error : ValidatedOptions.default
              }
            />
          </FormGroup>
        </Form>
      </Modal>
    </>
  );
};
