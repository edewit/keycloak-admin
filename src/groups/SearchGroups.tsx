import React from "react";
import {
  Button,
  ButtonVariant,
  Form,
  InputGroup,
  PageSection,
  PageSectionVariants,
  Text,
  TextContent,
  TextInput,
} from "@patternfly/react-core";
import { useTranslation } from "react-i18next";
import { SearchIcon } from "@patternfly/react-icons";

export const SearchGroups = () => {
  const { t } = useTranslation("groups");
  return (
    <>
      <PageSection variant={PageSectionVariants.light}>
        <TextContent className="pf-u-mr-sm">
          <Text component="h1">{t("searchForGroups")}</Text>
        </TextContent>
        <Form className="pf-u-mt-sm">
          <InputGroup>
            <TextInput
              name="search"
              id="group-search"
              type="search"
              aria-label={t("search")}
              placeholder={t("searchGroups")}
            />
            <Button variant={ButtonVariant.control} aria-label={t("search")}>
              <SearchIcon />
            </Button>
          </InputGroup>
        </Form>
      </PageSection>
    </>
  );
};
