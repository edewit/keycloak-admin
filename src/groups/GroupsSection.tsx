import React, { useContext, useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { HttpClientContext } from "../context/http-service/HttpClientContext";
import { GroupsList } from "./GroupsList";
import { GroupRepresentation } from "./models/groups";
import {
  ServerGroupsArrayRepresentation,
  ServerGroupMembersRepresentation,
} from "./models/server-info";
import { TableToolbar } from "../components/table-toolbar/TableToolbar";
import {
  Button,
  Divider,
  Dropdown,
  DropdownItem,
  KebabToggle,
  PageSection,
  PageSectionVariants,
  Spinner,
  Title,
  TitleSizes,
  ToolbarItem,
} from "@patternfly/react-core";
import "./GroupsSection.css";

export const GroupsSection = () => {
  const { t } = useTranslation("groups");
  const httpClient = useContext(HttpClientContext)!;
  const [rawData, setRawData] = useState<{ [key: string]: any }[]>();
  const [filteredData, setFilteredData] = useState<object[]>();
  const [isKebabOpen, setIsKebabOpen] = useState(false);
  const columnID: keyof GroupRepresentation = "id";
  const membersLength: keyof GroupRepresentation = "membersLength";
  const columnGroupName: keyof GroupRepresentation = "name";

  useEffect(() => {
    (async () => {
      const groups = await httpClient.doGet<ServerGroupsArrayRepresentation[]>(
        "/admin/realms/master/groups"
      );
      const groupsData = groups.data!;

      const getMembers = async (id: number) => {
        const response = await httpClient.doGet<
          ServerGroupMembersRepresentation[]
        >(`/admin/realms/master/groups/${id}/members`);
        const responseData = response.data!;
        return responseData.length;
      };

      const memberPromises = groupsData.map((group: { [key: string]: any }) =>
        getMembers(group[columnID])
      );
      const memberData = await Promise.all(memberPromises);
      const updatedObject = groupsData.map(
        (group: { [key: string]: any }, i: number) => {
          const object = Object.assign({}, group);
          object[membersLength] = memberData[i];
          return object;
        }
      );

      setRawData(updatedObject);
    })();
  }, []);

  // Filter groups
  const filterGroups = (newInput: string) => {
    const localRowData = rawData!.filter((obj: { [key: string]: string }) => {
      const groupName = obj[columnGroupName];
      return groupName.toLowerCase().includes(newInput.toLowerCase());
    });
    setFilteredData(localRowData);
  };

  // Kebab delete action
  const onKebabToggle = (isOpen: boolean) => {
    setIsKebabOpen(isOpen);
  };

  const onKebabSelect = () => {
    setIsKebabOpen(!isKebabOpen);
  };

  return (
    <React.Fragment>
      <PageSection variant={PageSectionVariants.light}>
        <Title headingLevel="h3" size={TitleSizes["2xl"]}>
          {t("groups")}
        </Title>
      </PageSection>
      <Divider />
      <PageSection variant={PageSectionVariants.light}>
        {!rawData && (
          <div style={{ textAlign: "center" }}>
            <Spinner />
          </div>
        )}
        {rawData && (
          <TableToolbar
            inputGroupName="groupsToolbarTextInput"
            inputGroupPlaceholder="Search groups"
            inputGroupOnChange={filterGroups}
            toolbarItem={
              <>
                <ToolbarItem>
                  <Button variant="primary">{t("createGroup")}</Button>
                </ToolbarItem>
                <ToolbarItem>
                  <Dropdown
                    onSelect={onKebabSelect}
                    toggle={<KebabToggle onToggle={onKebabToggle} />}
                    isOpen={isKebabOpen}
                    isPlain
                    dropdownItems={[
                      <DropdownItem key="action" component="button">
                        {t("delete")}
                      </DropdownItem>,
                    ]}
                  />
                </ToolbarItem>
              </>
            }
          >
            <GroupsList list={filteredData || rawData} />
          </TableToolbar>
        )}
      </PageSection>
    </React.Fragment>
  );
};
