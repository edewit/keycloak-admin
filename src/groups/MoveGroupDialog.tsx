import React, { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useErrorHandler } from "react-error-boundary";
import {
  Breadcrumb,
  BreadcrumbItem,
  Button,
  ButtonVariant,
  DataList,
  DataListAction,
  DataListCell,
  DataListItem,
  DataListItemCells,
  DataListItemRow,
  InputGroup,
  Modal,
  ModalVariant,
  TextInput,
  Toolbar,
  ToolbarContent,
  ToolbarItem,
} from "@patternfly/react-core";
import { AngleRightIcon, SearchIcon } from "@patternfly/react-icons";

import GroupRepresentation from "keycloak-admin/lib/defs/groupRepresentation";
import { asyncStateFetch, useAdminClient } from "../context/auth/AdminClient";
import { ListEmptyState } from "../components/list-empty-state/ListEmptyState";
import { GroupQuery } from "keycloak-admin/lib/resources/groups";

type MoveGroupDialogProps = {
  group: GroupRepresentation;
  onClose: () => void;
  onMove: (groupId: string | undefined) => void;
};

export const MoveGroupDialog = ({
  group,
  onClose,
  onMove,
}: MoveGroupDialogProps) => {
  const { t } = useTranslation("groups");

  const adminClient = useAdminClient();
  const errorHandler = useErrorHandler();

  const [navigation, setNavigation] = useState<GroupRepresentation[]>([]);
  const [groups, setGroups] = useState<GroupRepresentation[]>([]);
  const [filtered, setFiltered] = useState<GroupRepresentation[]>();
  const [searchValue, setSearchValue] = useState("");
  const [filter, setFilter] = useState("");

  const [id, setId] = useState<string>();
  const currentGroup = () => navigation[navigation.length - 1];

  const getLeafGroup = (
    groups: GroupRepresentation[]
  ): GroupRepresentation[] => {
    let result: GroupRepresentation[] = [];

    for (const group of groups) {
      if (group.subGroups?.length === 0) {
        result.push(group);
      } else {
        result = result.concat(getLeafGroup(group.subGroups!));
      }
    }

    return result;
  };

  const handleKeyDown = (e: { key: string }) => {
    if (e.key === "Enter") {
      setFilter(searchValue);
    }
  };

  useEffect(
    () =>
      asyncStateFetch(
        async () => {
          if (id) {
            const group = await adminClient.groups.findOne({ id });
            return { group, groups: group.subGroups! };
          } else {
            const params: GroupQuery = {};
            if (filter !== "") {
              params.search = filter;
            }
            return { groups: await adminClient.groups.find(params) };
          }
        },
        ({ group: selectedGroup, groups }) => {
          if (selectedGroup) setNavigation([...navigation, selectedGroup]);
          if (filter) {
            setFiltered(getLeafGroup(groups));
          } else {
            setGroups(groups.filter((g) => g.id !== group.id));
          }
        },
        errorHandler
      ),
    [id, filter]
  );

  return (
    <Modal
      variant={ModalVariant.large}
      title={t("moveToGroup", {
        group1: group.name,
        group2: currentGroup() ? currentGroup().name : filter ? "" : t("root"),
      })}
      isOpen={true}
      onClose={onClose}
      actions={[
        <Button
          data-testid="moveGroup"
          key="confirm"
          variant="primary"
          form="group-form"
          isDisabled={!!filter}
          onClick={() => onMove(currentGroup()?.id)}
        >
          {t("moveHere")}
        </Button>,
        <Button
          data-testid="moveCancel"
          key="cancel"
          variant="link"
          onClick={onClose}
        >
          {t("common:cancel")}
        </Button>,
      ]}
    >
      <Toolbar>
        <ToolbarContent>
          <ToolbarItem>
            <InputGroup>
              <TextInput
                type="search"
                aria-label={t("common:search")}
                placeholder={t("searchForGroups")}
                onChange={(value) => setSearchValue(value)}
                onKeyDown={handleKeyDown}
              />
              <Button
                variant={ButtonVariant.control}
                aria-label={t("common:search")}
                onClick={() => setFilter(searchValue)}
              >
                <SearchIcon />
              </Button>
            </InputGroup>
          </ToolbarItem>
        </ToolbarContent>
      </Toolbar>
      {!filter && (
        <Breadcrumb>
          <BreadcrumbItem key="home">
            <Button
              variant="link"
              onClick={() => {
                setId(undefined);
                setNavigation([]);
              }}
            >
              {t("groups")}
            </Button>
          </BreadcrumbItem>
          {navigation.map((group, i) => (
            <BreadcrumbItem key={i}>
              {navigation.length - 1 !== i && (
                <Button
                  variant="link"
                  onClick={() => {
                    setId(group.id);
                    setNavigation([...navigation].slice(0, i));
                  }}
                >
                  {group.name}
                </Button>
              )}
              {navigation.length - 1 === i && <>{group.name}</>}
            </BreadcrumbItem>
          ))}
        </Breadcrumb>
      )}

      <DataList
        onSelectDataListItem={(value) => {
          setFilter("");
          setFiltered(undefined);
          setId(value);
        }}
        aria-label={t("groups")}
        isCompact
      >
        {(filtered || groups).map((group) => (
          <DataListItem
            aria-labelledby={group.name}
            key={group.id}
            id={group.id}
          >
            <DataListItemRow data-testid={group.name}>
              <DataListItemCells
                dataListCells={[
                  <DataListCell key={`name-${group.id}`}>
                    <>
                      {filtered && filtered?.length !== 0
                        ? group.path
                        : group.name}
                    </>
                  </DataListCell>,
                ]}
              />
              <DataListAction
                aria-labelledby={`select-${group.name}`}
                id={`select-${group.name}`}
                aria-label={t("groupName")}
                isPlainButtonAction
              >
                <Button isDisabled variant="link">
                  <AngleRightIcon />
                </Button>
              </DataListAction>
            </DataListItemRow>
          </DataListItem>
        ))}
        {(filtered || groups).length === 0 && (
          <ListEmptyState
            hasIcon={false}
            message={t("moveGroupEmpty")}
            instructions={t("moveGroupEmptyInstructions")}
          />
        )}
      </DataList>
    </Modal>
  );
};
