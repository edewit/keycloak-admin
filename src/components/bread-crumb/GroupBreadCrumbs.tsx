import React, { useEffect } from "react";
import { Link, useHistory } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { Breadcrumb, BreadcrumbItem } from "@patternfly/react-core";

import { useSubGroups } from "../../groups/GroupsSection";

export const GroupBreadCrumbs = () => {
  const { t } = useTranslation();
  const { clear, remove, subGroups } = useSubGroups();

  const history = useHistory();

  useEffect(() => {
    return history.listen(({ pathname }) => {
      if (!pathname.startsWith("/groups") || pathname === "/groups") {
        clear();
      }
    });
  }, [history]);

  return (
    <>
      {subGroups.length !== 0 && (
        <Breadcrumb>
          <BreadcrumbItem key="home">
            <Link to="/groups">{t("groups")}</Link>
          </BreadcrumbItem>
          {subGroups.map((group, i) => (
            <BreadcrumbItem key={i} isActive={subGroups.length - 1 === i}>
              {subGroups.length - 1 !== i && (
                <Link
                  to={location.pathname.substr(
                    0,
                    location.pathname.indexOf(group.id!) + group.id!.length
                  )}
                  onClick={() => remove(group)}
                >
                  {group.name}
                </Link>
              )}
              {subGroups.length - 1 === i && <>{group.name}</>}
            </BreadcrumbItem>
          ))}
        </Breadcrumb>
      )}
    </>
  );
};
