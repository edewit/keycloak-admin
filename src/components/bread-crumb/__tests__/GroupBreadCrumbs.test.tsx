import React, { useEffect } from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";

import { GroupBreadCrumbs } from "../GroupBreadCrumbs";
import { SubGroups, useSubGroups } from "../../../groups/GroupsSection";

const GroupCrumbs = () => {
  const { addSubGroup } = useSubGroups();
  useEffect(() => {
    addSubGroup({ id: "123", name: "active group" });
  }, []);

  return <GroupBreadCrumbs />;
};

describe("Group BreadCrumbs tests", () => {
  it("couple of crumbs", () => {
    const crumbs = mount(
      <MemoryRouter initialEntries={["/groups"]}>
        <SubGroups>
          <GroupCrumbs />
        </SubGroups>
      </MemoryRouter>
    );

    expect(crumbs.find(GroupCrumbs)).toMatchSnapshot();
  });
});
