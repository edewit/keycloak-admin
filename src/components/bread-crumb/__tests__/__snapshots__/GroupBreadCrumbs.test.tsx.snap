// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group BreadCrumbs tests couple of crumbs 1`] = `
<GroupCrumbs>
  <GroupBreadCrumbs>
    <Breadcrumb>
      <nav
        aria-label="Breadcrumb"
        className="pf-c-breadcrumb"
        data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
        data-ouia-component-type="PF4/Breadcrumb"
        data-ouia-safe={true}
      >
        <ol
          className="pf-c-breadcrumb__list"
        >
          <BreadcrumbItem
            isActive={true}
            key=".$0"
            showDivider={false}
          >
            <li
              className="pf-c-breadcrumb__item"
            >
              active group
            </li>
          </BreadcrumbItem>
        </ol>
      </nav>
    </Breadcrumb>
  </GroupBreadCrumbs>
</GroupCrumbs>
`;
